Domain decomposition

mcapassox opened the pull request:
Merging of domain decomposition with k-equation. Adaptivity is now parallelised, domain decomposition is done in the SolEdge-HDG code instead of matlab and only one solution is saved instead of one per process. loading of the solution has been modified accordingly.

mcapassox committed:
beautified adaptivity

mcapassox committed:
beautified Definitions

mcapassox committed:
beautified Definitions

mcapassox committed:
Beautified HDG folder except HDG_CompJac

mcapassox committed:
Beautified Inout folder

mcapassox committed:
Beautified LinearAlgrebra

mcapassox committed:
Beautified Main

mcapassox committed:
Ugly fix for the external faces flipface. Solved difference between serial and parallel version.

mcapassox committed:
added forgotten deallocations in globals

mcapassox committed:
initialised pointers to NULL in types.f90

mcapassox committed:
added fix for analytical_solution used to initialise or to apply dirichlet boundary conditions

mcapassox committed:
small fixes on MACROS

mcapassox committed:
delta becomes an interger and WEST supports again dirichlet boundary conditions

mcapassox committed:
a couple of arrays where not initialised

mcapassox committed:
initialised pointers to NULL

mcapassox committed:
initialisation of some arrays

mcapassox committed:
initialisation of few arrays

mcapassox committed:
small fixes

mcapassox committed:
small fixes on types

mcapassox committed:
changed MPI_DOUBLE_PRECISION TO MPI_REAL8 and added a few routines to gather parallel solution into serial

mcapassox committed:
small fixes on types and reals

mcapassox committed:
small fixes on types and reals

mcapassox committed:
small fixes on tollerances and reals

mcapassox committed:
modified MPI_DOUBLE_PRECISION to MPI_REAL8

mcapassox committed:
modified MPI_DOUBLE_PRECISION to MPI_REAL8

mcapassox committed:
added forgotten deallocations

mcapassox committed:
removed useless attributes to solution

mcapassox committed:
did some cleaning

mcapassox committed:
alignement stuff

mcapassox committed:
added nullification of pointers and forgotten deallocations

mcapassox committed:
addition of this file to clean the MAIN of variables declarations and functions

mcapassox committed:
cleaned MAIN of variables' declaration and subroutine, they are now in Main_utils

mcapassox committed:
added deallocation of Mesh%flag_elems_sc

mcapassox committed:
just changed formatting

mcapassox committed:
just changed formatting

mcapassox committed:
added subrotuine compute_error_oscillations

mcapassox committed:
minor changes

mcapassox committed:
changed formatting

mcapassox committed:
modified gmsh%open to accept adapt%geometry_path

mcapassox committed:
addition of add_initial_perturbation

mcapassox committed:
added space

mcapassox committed:
added reading of geometry_path

mcapassox committed:
modified name from load_mesh to load_mesh_serial_h5

mcapassox committed:
added function read_splines

mcapassox committed:
macros were placed in wrong places

mcapassox committed:
formatting and addition of initialize_puff

mcapassox committed:
added Main_utils to Makefile

mcapassox committed:
remouved temporary files

mcapassox committed:
modified call to load_mesh_h5

mcapassox committed:
added some output printing messages

mcapassox committed:
elemsize was not freed. now it is. it also became a pointer and not an allocatable.

mcapassox committed:
beautified and added a few entries for structure mesh in parallel

mcapassox committed:
fixed typo

mcapassox committed:
removed ugly fix on flifaces over external faces

mcapassox committed:
only process 0 prints dialogs

mcapassox committed:
fixed wrong deallocation of Xel and Xfl

mcapassox committed:
modified computeNodalConnectivity and computeElementSize so that they can handle output arguments, fixed bug on computePumpArea ad computePuffArea and only process 0 prints output

mcapassox committed:
only process 0 prints output and fixed boundary connectivity reading issue in case PUFF/PUMP/IN/OUT are not in the same order on the gmsh file as in the fortran code

mcapassox committed:
removed some unused variables

mcapassox committed:
modifed some tolerances TODO: to check again when merging

mcapassox committed:
beautified and small fixes

mcapassox committed:
added file domain_decomposition

mcapassox committed:
didn't really change anything...

mcapassox committed:
added a bunch of routines to gather solutions, mesh, magnetic field and general vectors in 1 and 2D over the poloidal processes.

mcapassox committed:
modified load_mesh, load_mesh_from_solution and saving solution so that one single solution file is saved even for parallel simulations. Now the solution saves groups mesh, solution, magnetic and simulation_parameters.

mcapassox committed:
added the file to make the main more readable and added solution_decomposition to decompose a serial solution over the processes. Memory is better managed with corrections on how arrays and pointers are deallocated/nullified

mcapassox committed:
added some printing and fixed undeallocated arrays. added compute_error_on_vertices_root for parallel enviroment.

mcapassox committed:
parallelised estimator-indicator

mcapassox committed:
parallelised estimator

mcapassox committed:
parallelised indicator

mcapassox committed:
added call to solution_decomposition and domain_decomposition

mcapassox committed:
added domain_decomposition to the files to be compiled

mcapassox committed:
corrected loading solution and domain/solution decomposition for parallel and serial versions

mcapassox committed:
added X_P1 to solution saving

mcapassox committed:
modified formatting

mcapassox committed:
resolved merge conflicts

mcapassox committed:
use solve_pastix was not needed

mcapassox committed:
only imported the necessary functions from PETSC

mcapassox committed:
only imported the necessary functions from PETSC

mcapassox committed:
only imported the necessary functions from solve_pastix

mcapassox committed:
only imported the necessary functions from solve_petsc and removed USE PASTIX/PSBLAS

mcapassox committed:
formatting

mcapassox committed:
hwloc was not linked against PASTIX

wave46 commented on lib/Make.inc/arch.make:
```diff
@@ -64,8 +64,8 @@ PASTIX=$(LIB_YES)
-#PETSC=$(LIB_YES)
-PETSC=$(LIB_NO)
+PETSC=$(LIB_YES)
+#PETSC=$(LIB_NO)
```
@mcapassox 
```suggestion
#PETSC=$(LIB_YES)
PETSC=$(LIB_NO)
```

mcapassox committed:
Update lib/Make.inc/arch.make

Co-authored-by: wave46 <59137482+wave46@users.noreply.github.com>

wave46 commented on src/Models/NGammaTiTe/physics.f90:
```diff
@@ -1713,30 +1716,30 @@ SUBROUTINE compute_dsigmavnn_dU(U,res)
-      dti_dU(:) = dti_dU(:) * T0*2/3. /phys%Mref
-
+      dti_dU(:) = dti_dU(:) * T0*2./3. /phys%Mref
+      ti = T0*2./3. /phys%Mref * (U3/U1 - 0.5 *U2**2/U1**2)
```
```suggestion
```
no need to recalculate again

wave46 committed:
Remove unnecessary line in  src/Models/NGammaTiTe/physics.f90

mcapassox committed:
cleaned file dependencies called by USE

mcapassox committed:
cleaned file dependencies. parallel compilation should work for serial and parallel versions.

mcapassox committed:
Merge branch 'domain_decomposition' of ssh://github.com/wave46/MHDG into domain_decomposition
I need the global modifications on my local brunch

mcapassox committed:
added forgotten USE

mcapassox committed:
added if condition. in case petsc is used solve_petsc.o need to be linked to main_utils.

mcapassox committed:
well, the fix was wrong. So i fixed the fix

wave46 commented on src/MHDG.f90:
```diff
@@ -125,164 +41,60 @@ PROGRAM MHDG
   ierr = 1
-  CALL mesh_preprocess_serial(ierr)
+#ifdef PARALL
+  IF(switch%read_gmsh) THEN
```
```suggestion
  IF(switch%read_gmsh .or. switch%readMeshFromSol) THEN
```
Because now the parallel version of the code also provides "serial" output

wave46 committed:
Removing Mesh from private clause

It simply shouldn't be private and if it is, in parallel memory errors occur
maybe same should be done with phys

wave46 committed:
Update src/MHDG.f90

wave46 commented on src/Models/magnetic_field.f90:
```diff
@@ -32,13 +35,20 @@ SUBROUTINE initialize_magnetic_field
     END IF
   END SUBROUTINE initialize_magnetic_field
 
+  SUBROUTINE load_magnetic_field_Jtor
+    IF (switch%ME .EQV. .FALSE.) THEN
+       CALL load_magnetic_field()
+       IF (switch%ohmicsrc) CALL loadJtorMap()
+    END IF
+  ENDSUBROUTINE load_magnetic_field_Jtor
+
```
```suggestion
  SUBROUTINE load_magnetic_field_Jtor
      CALL load_magnetic_field()
      IF (switch%ohmicsrc) CALL loadJtorMap()
  ENDSUBROUTINE load_magnetic_field_Jtor

```

No matter this is moving equilibrium or not, we should update magnetic field following update/initialize of solution/mesh.
should be updated together with `initialize_solution` in `Main_utils.f90`

wave46 commented on src/Utils/Main_utils.f90:
```diff
@@ -0,0 +1,1079 @@
+       ! restart simulation: load solution from file (the name is given in argument)
+       CALL HDF5_load_solution(save_name)
+       ALLOCATE(sol%u_tilde0(SIZE(sol%u_tilde)))
+       ! Update magnetic field and, if ohmic src, Jtor also
+       ! In case of restart initialize the magnetic configuration to the previous one
+       IF (switch%ME .EQV. .TRUE.) THEN
+          CALL load_magnetic_field()
+          IF (switch%ohmicsrc) CALL loadJtorMap()
+       END IF
```
```suggestion
```
Magnetic field will be initialized by `load_magnetic_field_Jtor` from 'magnetic_field.f90'

wave46 committed:
Update src/Models/magnetic_field.f90

wave46 committed:
Update src/Utils/Main_utils.f90

wave46 commented on src/MHDG.f90:
```diff
@@ -381,433 +152,112 @@ PROGRAM MHDG
+        ! Check for NaN (should work with optimization flags)
+        CALL check_for_NaNs()
+
+        IF (adapt%adaptivity .AND. restart_adapt) THEN
+           CALL adaptivity()
```
```suggestion
        IF (switch%ME .EQV. .TRUE.) THEN
              time%it=time%it-1
           ENDIF
           CALL adaptivity()
           IF (switch%ME .EQV. .TRUE.) THEN
            time%it=time%it+1
           ENDIF
```

This is a more hotfix to correct the time counter for moving equilibrium magnetic field and jtor update. Should be a nicer way to do so

wave46 commented on src/MHDG.f90:
```diff
@@ -816,338 +266,200 @@ PROGRAM MHDG
 
-              Mesh%X = Mesh%X*phys%lscale
-              Mesh_prec%X = Mesh_prec%X*phys%lscale
+              ! call adaptivity precedure
+              CALL adaptivity()
```
```suggestion
              IF (switch%ME .EQV. .TRUE.) THEN
               time%it=time%it-1
              ENDIF
              ! call adaptivity precedure
              CALL adaptivity()
              IF (switch%ME .EQV. .TRUE.) THEN
               time%it=time%it-1
              ENDIF
```

Still more than a hotfix than solution

wave46 reviewed (changes requested):
@mcapassox can you take a look if this hotfix is reasonable?

wave46 commented on src/MPI_OMP/Communications.f90:
```diff
@@ -615,5 +618,601 @@ END SUBROUTINE set_permutations
+    IF(PRESENT(periodic_faces_glob)) DEALLOCATE(periodic_faces)
+  ENDSUBROUTINE gather_mesh
+
+  SUBROUTINE gather_solution(Mesh_in, Nnodesperelem, Nnodesperface, u_tilde_in, u_in, q_in, u_glob, u_tilde_glob, q_glob)
```
There is something with gathering or saving u_tilde, because, when restarting from other solution, u_tilde used in first NR iteration is definitely wrong, leading to incorrect solution of first NR iteration. As a brute force solution, `extractFaceSolution()` can be used after `HDF5_load_solution(save_name)`, but this is not acceptable solution. It is neded to check, how u_tilde is solved

wave46 replied:
If restarting from a serial solution (with either serial or parallel solver) everything is fine. So, indeed, u_tilde is not correctly saved and/or gathered in parallel version of the code.
@mcapassox will you take a look how to fix?

wave46 reviewed (changes requested):
@mcapassox, and I'll explore a bit deeper this

wave46 reviewed (changes requested):
Actually, [commit](https://github.com/wave46/MHDG/pull/1/commits/835d19490216875971dc1f7babf7e57d7c35347c) not enough, one still has to keep both switches as true in switch%read_gmsh and switch%readMeshFromSo

mcapassox committed:
fixed bug on mesh not decomposed over processes when restart simulation with loadMeshFromSol switch active

mcapassox committed:
fixed bug on mesh not decomposed over processes when restart simulation with loadMeshFromSol switch active and newmesh_n is now saved as global mesh

mcapassox committed:
unimportant merge

mcapassox committed:
just fixed indentation

mcapassox committed:
fixed bug on how Tb, extfaces, intfaces, boundary_flag and periodic_faces are gathered

wave46 committed:
Update src/MHDG.f90 to work in moving equilibrium mode

wave46 committed:
Update src/MHDG.f90 to work in moving equilibrium mode

wave46 committed:
hotfix, line dissappeared

wave46 committed:
Moving equilibrium solution load fix

This part of the code is not needed, since puff_exp is anyway loaded later from other file
Otherweise, puff_exp is not allocated, therefore it gives an error

wave46 committed:
Fix wrong neutral diffusion output

Neutral diffusion has been plotted adimensional

wave46 committed:
input of mesh geometry in the sample param files

updating param files since now path to geometry is not hardcoded

wave46 committed:
Typo correction

wave46 committed:
small cleaning of param files

wave46 commented on src/Adaptivity/adaptivity_estimator_indicator.f90:
```diff
@@ -101,94 +126,138 @@ SUBROUTINE adaptivity_indicator_estimator(mesh_name,thresh, param_adapt, count_a
+    !! error indicator based on the elemental oscillations
+#ifndef PARALL
+    CALL read_error(eps_plot, error_oscillation)
+    h_target = h  ! Start by assigning h to h_target
```
```suggestion
```

This line just erased data from estimator

wave46 committed:
Update src/Adaptivity/adaptivity_estimator_indicator.f90

wave46 committed:
Removing unnecessary reduction of neutral diffusion

It was not applied anyway

wave46 commented on src/Adaptivity/adaptivity_estimator_indicator.f90:
```diff
@@ -101,94 +126,137 @@ SUBROUTINE adaptivity_indicator_estimator(mesh_name,thresh, param_adapt, count_a
+       error_oscillation_root = 0.
+       CALL compute_error_on_vertices_root(error_oscillation_glob, vector_nodes_unique_glob, count_vec_glob, error_oscillation_root)
 
+       h_target = h_root  ! Start by assigning h to h_target
```
```suggestion
```
Overwrited the estimator data

wave46 committed:
Update src/Adaptivity/adaptivity_estimator_indicator.f90

wave46 committed:
Better for allocation

wave46 committed:
Removing read_gmsh from MHDG

We now run only from gmsh meshes and the mesh will be saved always in solution

Now either read_gmsh or readMeshFromSol should be true anyway

wave46 committed:
Removing reag_gmsh from Main_utils

We now run only from gmsh meshes and the mesh will be saved always in solution

Now either read_gmsh or readMeshFromSol should be true anyway

wave46 committed:
Remove another read_gmsh switch from MHDG

We now run only from gmsh meshes and the mesh will be saved always in solution

Now either read_gmsh or readMeshFromSol should be true anyway

wave46 committed:
Adding Error drop if not read gmsh mesh or from solution

wave46 committed:
Removing saveMeshSol from inout 1

now the mesh will be saved in solution by default

wave46 committed:
Removing rest savemeshsol from inout

wave46 committed:
Removing switch saveMeshSol

wave46 committed:
To safely deallocate mesh tuff

wave46 committed:
fix to gmsh reader for large orders

wave46 committed:
udpdating readme and support param files
